pkg/controller/compute/router.go

```diff
@@ -0,0 +1,181 @@
+	errRouterDeleteFailed  = "deletion of Router resource has failed"
+	errCheckRouterUpToDate = "cannot determine if GCP Router is up to date"
+)
+
+// SetupRouter adds a controller that reconciles Router managed
+// resources.
+func SetupRouter(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(v1beta1.RouterGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1beta1.Router{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.RouterGroupVersionKind),
+			managed.WithExternalConnecter(&routerConnector{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type routerConnector struct {
+	kube client.Client
+}
+
+func (c *routerConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := compute.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &routerExternal{Service: s, kube: c.kube, projectID: projectID}, nil
+}
+
+type routerExternal struct {
+	kube client.Client
+	*compute.Service
+	projectID string
+}
+
+func (c *routerExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Router)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotRouter)
+	}
+	observed, err := c.Routers.Get(c.projectID, cr.Spec.ForProvider.Region, meta.GetExternalName(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetRouter)
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	router.LateInitializeSpec(&cr.Spec.ForProvider, *observed)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errManagedRouterUpdate)
+		}
+	}
+
+	cr.Status.AtProvider = router.GenerateRouterObservation(*observed)
+
+	cr.Status.SetConditions(xpv1.Available())
+
+	u, err := router.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckRouterUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: u,
+	}, nil
+}
+
+func (c *routerExternal) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.Router)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotRouter)
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
```

Comment:
I believe we can just drop this line, covered by crossplane runtime managed reconciler: https://github.com/crossplane/crossplane-runtime/blob/ba474e81c62c8e92a810d471a9a04efa62245817/pkg/reconciler/managed/reconciler.go#L714